src/stan/mcmc/chains.hpp

```diff
@@ -292,6 +291,91 @@ namespace stan {
         return ess;
       }
 
+      /**
+       * Returns the effective sample size for the specified parameter
+       * across all kept samples.
+       *
+       * The implementation is close to the effective sample size
+       * description in BDA3 (p. 286-287).  See more details in Stan
+       * reference manual section "Effective Sample Size".
+       *
+       * Current implementation takes the minimum number of samples
+       * across chains as the number of samples per chain.
+       *
+       * @param std::vector<std::vector<double> > samples
+       * @return
+       */
+      double
+      effective_sample_size(const std::vector<std::vector<double> >& samples)
+        const {
+        int chains = samples.size();
+
+        // need to generalize to each jagged samples per chain
+        int n_samples = samples[0].size();
+        for (int chain = 1; chain < chains; chain++) {
+          n_samples = std::min(n_samples,
+                               static_cast<int>(samples[chain].size()));
+        }
+
+        std::vector<std::vector<double> > acov(chains);
+        for (int chain = 0; chain < chains; chain++) {
+          stan::math::autocovariance(samples[chain], acov[chain]);
+        }
+
+        std::vector<double> chain_mean(chains);
+        std::vector<double> chain_var(chains);
+        for (int chain = 0; chain < chains; chain++) {
```

Comment:
We prefer `++chain` to `chain++` when the pre-increment value isn't needed.  There are a bunch of these.